backends/dlc/scummvmcloud.cpp

```diff
@@ -0,0 +1,232 @@
+	DLCMan.refreshDLCList();
+}
+
+void ScummVMCloud::errorCallbackGetAllDLCs(Networking::ErrorResponse error) {
+	warning("JsonRequest Error - getAllDLCs");
+}
+
+void ScummVMCloud::getAllDLCs() {
+	Common::String url("https://scummvm-dlcs-default-rtdb.firebaseio.com/dlcs.json"); // temporary mock api
+	Networking::CurlJsonRequest *request = new Networking::CurlJsonRequest(
+		new Common::Callback<ScummVMCloud, Networking::JsonResponse>(this, &ScummVMCloud::jsonCallbackGetAllDLCs), 
+		new Common::Callback<ScummVMCloud, Networking::ErrorResponse>(this, &ScummVMCloud::errorCallbackGetAllDLCs), 
+		url);
+
+	request->execute();
+}
+
+void ScummVMCloud::downloadFileCallback(Networking::DataResponse r) {
+	Networking::SessionFileResponse *response = static_cast<Networking::SessionFileResponse *>(r.value);
+	DLCMan._currentDownloadedSize += response->len;
+	if (DLCMan._interruptCurrentDownload) {
+		_rq->close();
+		DLCMan._interruptCurrentDownload = false;
+		// delete the download cache (the incomplete .zip)
+		Common::Path relativeFilePath = Common::Path(DLCMan._queuedDownloadTasks.front()->id);
+		removeCacheFile(relativeFilePath);
+
+		DLCMan._queuedDownloadTasks.front()->state = DLCDesc::kCancelled;
+		DLCMan.refreshDLCList();
+
+		// handle next download
+		DLCMan._queuedDownloadTasks.pop();
+		DLCMan._dlcsInProgress.remove_at(0);
+		DLCMan.processDownloadQueue();
+	}
+	if (response->eos) {
+		DLC::DLCDesc *dlc = DLCMan._queuedDownloadTasks.front();
+		debug(1, "Downloaded: %s", dlc->name.c_str());
+
+		_rq->close(); // delete request
+
+		Common::Path relativeFilePath = Common::Path(dlc->id);
+
+		// extract the downloaded zip
+		Common::String gameDir = Common::punycode_encodefilename(dlc->name);
+		Common::Path destPath = Common::Path(ConfMan.get("dlcspath")).appendComponent(gameDir);
+		Common::Error error = extractZip(relativeFilePath, destPath);
+
+		// remove cache (the downloaded .zip)
+		removeCacheFile(relativeFilePath);
+
+		if (error.getCode() == Common::kNoError) {
+			// add downloaded game entry in scummvm configuration file
+			addEntryToConfig(destPath);
+			dlc->state = DLCDesc::kDownloaded;
+			DLCMan._errorText = "";
+		} else {
+			// if there is any error in extraction
+			dlc->state = DLCDesc::kErrorDownloading;
+			DLCMan._errorText = error.getDesc();
+		}
+
+		DLCMan.refreshDLCList();
+
+		// handle next download
+		DLCMan._queuedDownloadTasks.pop();
+		DLCMan._dlcsInProgress.remove_at(0);
+		DLCMan.processDownloadQueue();
+	}
+}
+
+void ScummVMCloud::errorCallback(Networking::ErrorResponse error) {
+	// error downloading - start next download in queue
+	DLCMan._queuedDownloadTasks.front()->state = DLCDesc::kErrorDownloading;
+	DLCMan._queuedDownloadTasks.pop();
+	DLCMan._dlcsInProgress.remove_at(0);
+	DLCMan.processDownloadQueue();
+}
+
+void ScummVMCloud::startDownloadAsync(const Common::String &id, const Common::String &url) {
+	Common::String localFile = normalizePath(ConfMan.get("dlcspath") + "/" + id, '/');
```

Comment:
You need to expose `dlcpath` in the GUI, Options -> Paths

Reply:
I have added it. There is, however, a bit of hole/space when `USE_DLC` is not defined (possibly due to spacing = 10 in layout element inside .stx). I have added the layout (containing widgets) for dlcpath at the end for consistent spacing between the buttons for different paths. The next widget below will have some space before.

For the record, in the launcher, there is no hole/space for the DLC button if `USE_DLC` is not defined.